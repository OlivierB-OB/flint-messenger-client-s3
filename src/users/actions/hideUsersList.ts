import { IToggleUserListAction, TOGGLE_USER_LIST } from '../types';

export function hideUsersList(): IToggleUserListAction {
  return {
    type: TOGGLE_USER_LIST,
    data: false,
  };
}
