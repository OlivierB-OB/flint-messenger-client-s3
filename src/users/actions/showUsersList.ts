import { IToggleUserListAction, TOGGLE_USER_LIST } from '../types';

export function showUsersList(): IToggleUserListAction {
  return {
    type: TOGGLE_USER_LIST,
    data: true,
  };
}
