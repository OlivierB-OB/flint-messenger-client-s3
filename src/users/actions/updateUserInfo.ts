import { IUpdateUserInfoAction, UPDATE_USER_INFO, IUserInfo } from '../types';

export function updateUserInfo(info: IUserInfo[]): IUpdateUserInfoAction {
  return {
    type: UPDATE_USER_INFO,
    data: info,
  };
}
