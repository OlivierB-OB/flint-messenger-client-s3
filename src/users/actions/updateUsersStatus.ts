import { IUpdateUsersStatusAction, UPDATE_USERS_STATUS, IUsersStateStatus } from '../types';

export function updateUsersStatus(status: IUsersStateStatus): IUpdateUsersStatusAction {
  return {
    type: UPDATE_USERS_STATUS,
    status,
  };
}
