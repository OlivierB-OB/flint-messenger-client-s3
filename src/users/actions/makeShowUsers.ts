import { batch } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { updateDrawerContent } from '../../layout/actions/updateDrawerContent';
import { showDrawer } from '../../layout/actions/showDrawer';

export function makeShowUsers() {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    batch(() => {
      dispatch(updateDrawerContent('contacts'));
      dispatch(showDrawer());
    });
  };
}
