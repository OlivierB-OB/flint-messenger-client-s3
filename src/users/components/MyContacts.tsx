import Box from '@material-ui/core/Box';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Drawer from '@material-ui/core/Drawer';
import React from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { makeCreateConversation } from '../../conversations/actions/makeCreateConversation';
import { hideUsersList } from '../actions/hideUsersList';
import { IUserInfo, IUsersStateStatus } from '../types';
import { ContactListItem } from './ContactListItem';
import { Loading } from '../../layout/components/Loading';

export interface IContactListProps {
  status: IUsersStateStatus;
  show: boolean;
  list: IUserInfo[];
  hideContactList: () => void;
  createConversation: (targetId: string) => void;
}

export function ContactList(props: IContactListProps) {
  const { status, list, show, hideContactList, createConversation } = props;
  const loading = status === 'unavailable' ? <Loading /> : null;
  return (
    <Drawer anchor="left" open={show} onClose={hideContactList}>
      <Box style={{ minWidth: '300px' }}>
        {loading}
        <List>
          {loading || list.length ? null : (
            <ListItem>
              <ListItemText primary="no contact available..." />
            </ListItem>
          )}
          {list.map((user) => (
            <ListItem button onClick={() => createConversation(user.uid)} key={user.uid}>
              <ContactListItem info={user} />
            </ListItem>
          ))}
        </List>
      </Box>
    </Drawer>
  );
}

const mapStateToProps = ({ users }: IAppState) => ({
  status: users.status,
  show: users.show,
  list: users.list,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  hideContactList: () => dispatch(hideUsersList()),
  createConversation: (targetId: string) => dispatch(makeCreateConversation(targetId)),
});

export const MyContacts = connect(mapStateToProps, mapDispatchToProps)(ContactList);
