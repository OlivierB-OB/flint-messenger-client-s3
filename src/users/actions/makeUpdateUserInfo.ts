import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { IUserInfo } from '../types';
import { updateIdentity } from '../../identity/actions/updateIdentity';
import { updateUserInfo } from './updateUserInfo';

export const makeUpdateUserInfo = action((infos: IUserInfo[]) => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    const myId = getState().identity.info?._id;
    const myInfo = infos.find(({ _id }) => _id === myId);
    if (myInfo) dispatch(updateIdentity(myInfo));
    dispatch(updateUserInfo(infos.filter(({ _id }) => _id !== myId)));
  };
});
