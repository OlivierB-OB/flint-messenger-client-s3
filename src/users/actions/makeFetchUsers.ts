import { batch } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { updateUsersStatus } from './updateUsersStatus';
import { defaultUsersState } from '../cases/defaultUsersState';
import { updateUserInfo } from './updateUserInfo';

const SLEEP = (delay: number) => new Promise((resolve) => setTimeout(resolve, delay));

export function makeFetchUsers() {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    dispatch(updateUsersStatus('unavailable'));

    // FIXME HTTP call goes here
    await SLEEP(3000);
    const { list } = defaultUsersState;

    if (!list) dispatch(updateUsersStatus('unavailable'));
    else {
      batch(() => {
        dispatch(updateUsersStatus('ready'));
        for (const user of list) dispatch(updateUserInfo(user));
      });
    }
  };
}
