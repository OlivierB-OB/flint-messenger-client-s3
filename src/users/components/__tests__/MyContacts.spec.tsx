import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { MyContacts } from '../MyContacts';
import { makeCreateConversation } from '../../../conversations/actions/makeCreateConversation';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('MyContacts', () => {
  let Connected: any = MyContacts;
  let spyCreateConversation: jest.Mock;
  let restoreCreateConversation: () => void;

  beforeEach(() => {
    [spyCreateConversation, restoreCreateConversation] = makeCreateConversation.mock(jest.fn());
  });

  afterEach(() => {
    restoreCreateConversation();
  });

  it('should provide the status, user list and a createConversation function to UserList', async () => {
    const store = mockStore({
      users: {
        status: 'the status',
        list: 'the user list',
      },
    });
    const component = shallow(<Connected store={store} />).dive();
    expect(component).toMatchSnapshot();
    (component.prop('createConversation') as any)('123');
    expect(spyCreateConversation).toHaveBeenCalledWith('123');
  });
});
