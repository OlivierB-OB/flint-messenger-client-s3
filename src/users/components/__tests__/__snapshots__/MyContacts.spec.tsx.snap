// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyContacts should provide the status, user list and a createConversation function to UserList 1`] = `
<ContactList
  addToConversation={[Function]}
  createConversation={[Function]}
  list="the user list"
  status="the status"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
